### ISSUE TITLE
Load Through Instead of Patch

### ISSUE THREAD

**Issue by Mpstark** [Mpstark/BattleTechModLoader#4]
CptMoore posted on ModTek's issues:

> Create BattleTechLoader.exe as a substitute for starting BattleTech.exe and that automatically patches any un-patched assemblies before starting BattleTech.exe automatically. Bascially a nice wrapper around BTML for users. Thats why Skyrim is doing.

**Comment by Mpstark** [comment]
This is a straight improvement to the current patch-in-place behavior.

**Comment by Tyler-IN** [comment]
Do you actually want to do this?

I don't think you should make a BattleTech.exe substitute or loader, but you could merge in a module initializer and pre-empt and delay the Unity startup, display your own GUI, and ultimately do the same thing...

I think the patch-in-place behavior can be more flexible and simpler in some ways than pre-runtime patching. Having both facilities available would be great though. Similar to "core mods" that have to modify the jar vs regular "mods" that are loaded by the mod loader in Minecraft.

**Comment by CptMoore** [comment]
> I think the patch-in-place behavior can be more flexible and simpler in some ways than pre-runtime patching.

What do you mean, what exactly would be simpler?

**Comment by Mpstark** [comment]
If the the launcher still patches the files on the hard disk, there is zero reason to change the fundamental nature of BTML. The only reason to change the fundamental nature of BTML is to change how the game is injected and I believe that injection through a launch-through process that doesn't touch the game files is basically the only way to improve the current behavior.

As to adding in wizbang features like checking versions -- straight no from me. Things that are complicated break and foundational parts of how mods *work* should not break, therefor BTML should not be complicated.

BTML also shouldn't be moved from the `\managed\` folder because it uses the games libraries to perform the patch, excepting Mono.Cecil which is included in the injector assembly. Since *the game is the only part a users setup* that we can rely on, this makes sense to me. Using the same libraries as the game reduces friction and therefor reduces complication.

Ultimately, I very much like the core choices that I've made so far for both ModTek and BTML. I want this project to stay working for as long as people are still playing the game -- surely you can understand the thought process of keeping everything simple.

**Comment by CptMoore** [comment]
> If the the launcher still patches the files on the hard disk, there is zero reason to change the fundamental nature of BTML.

That was not what had in mind for the launcher, the launcher is for automating patching after game updates, so improving the user experience by just clicking an exe called BTMLLauncher .

But you ask for additional features, like the ability to start the native experience without mods. That can be achieved by making the launcher communicate with BTMLoader and telling it to actually load mods. Patching is done regardless.

Making it patch only in memory would be another neat feature, but thats yet outside of my experiences with C#/C++ and unity.

Please tell me if you are still against a BTMLLauncher that just automates patches,  because that is what I was about to implement.

> surely you can understand the thought process of keeping everything simple at the foundational level.

In what is simple I always have to defer to your judgement since you are obviously the one who has to maintain this and coordinate all effort. My suggestions are often from myself needing a feature and seeing features available in other modded projects. Sometimes its from the view of a modder and sometimes from a gamers point of view. I'm sorry if I overload you with questions and demand answers all the time, but from my point whats simple for you is hard for me to guess.

**Comment by Mpstark** [comment]
There is a difference between building a tool *over* BTML and changing how it works. If you want to build a tool that depends on BTML or a tool that installs it, that's fine by me. It'd just be it's own thing, not be part of BTML if that makes any sense.

**Comment by CptMoore** [comment]
An installer is also needed. Would you mind if I create separate projects within the BTML sln and add those to this repo for the launcher and installer?

The installer might take some time, my knowledge about this is kinda old (~10 years ago with NSIS).

Why put the launcher and installer in the same sln? Well release packaging, it should be one click to create the btml loader, btml launcher and the btml installer.

Or dont you want any code of such in this repo? You know my stance, if it is smaller than the linux kernel put everything related in the same repo. Different artifacts are managed using different projects under one single sln.

**Comment by CptMoore** [comment]
see https://github.com/Qwiso/QModManager , they dont have a loader but and installer, they ask just to reinstall the installer to setup everything again

they seem to have the loader and the dependency management (modtek) as the same app

**Comment by Mpstark** [comment]
Look, I'm going to be real with you -- I like the way that it is. Small and self-contained. And I feel that I have clearly stated this and tried to be clear about what improvements that I think _would_ be valuable. The constant barrage of insinuations that I'm wrong about choices I made is starting to wear on me as well -- I understand that you would make different choices then me. Hell, I might have made different decisions three years ago and almost certainly will make different ones in the future.

I do not feel as if a launcher that still patches game files or an installer bring any *real* benefit to the project as it is now, in the very early days. Perhaps when BattleTech modding is significantly more mature and the codebase has been tested over some amount of time, then it might be good to have an installer/launcher. Perhaps when HBS is not patching the game every week.

Ultimately, I cannot prevent you from spending your time doing these things, even if I think that they are not well spent. I can tell you that I probably won't be accepting a pull request that adds them at least for some time.

**Comment by CptMoore** [comment]
> The constant barrage of insinuations that I'm wrong about choices I made is starting to wear on me as well

Since you feel that way, then I will cease doing that.

**Comment by Mpstark** [comment]
You've made lots of suggestions that I've followed through on, and I don't think you've made any *bad* or *incorrect* suggestions. It's just that when it's the same suggestion made multiple times after I've tried to explain my reasoning behind making it.

I don't know, I could just be sensitive or hard to work with and just need to get over it. Or maybe I need to say 'no' or 'not right now' better, I don't know. I'm a work-in-progress just as much as the project is.
